### PR TITLE
GetPeers() random order

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: p2p/peers.go [hunk 1]
```diff
@@ -40,8 +41,24 @@ func (pi PeersImpl) Close() {
 
 func (pi PeersImpl) GetPeers() []Peer {
 	peers := pi.snapshot.Load().([]Peer)
-	pi.With().Info("now connected", log.Int("n_peers", len(peers)))
-	return peers
+	cpy := make([]Peer, len(peers))
```
Comment (y0sher) [inline p2p/peers.go:44]: why do you copy? isn't snapshot loading always returns a copy? Until now we would've just returned peers.

Comment (almogdepaz) [reply p2p/peers.go:44]: because its a slice it returns a pointer so the copy is necessary

File: p2p/peers.go [hunk 2]
```diff
@@ -40,8 +41,24 @@ func (pi PeersImpl) Close() {
...
+	// We start at the end of the slice, inserting our random
+	// values one at a time.
+	for n := len(vals); len(vals) > 1 && n > 0; n-- {
+		randIndex := rand.Intn(len(vals))
+		// We swap the value at index n-1 and the random index
+		// to move our randomly chosen value to the end of the
+		// slice, and to move the value that was at n-1 into our
+		// unshuffled portion of the slice.
+		vals[n-1], vals[randIndex] = vals[randIndex], vals[n-1]
+	}
```
Comment (noamnelke) [inline p2p/peers.go]: I suggest using the built-in `rand.Shuffle` like here:
https://github.com/spacemeshos/go-spacemesh/blob/eaa676312044fa25911ded4961a6f441647a3192/mesh/mesh.go#L186-L189
It does more or less the same thing, with fewer opportunities to make mistakes.

**Review by y0sher — APPROVED** [review]

**Review by noamnelke — CHANGES_REQUESTED** [review]

File: p2p/peers.go (hunk 2 above)
Comment (noamnelke) [reply p2p/peers.go]: I was actually curious about a specific deviation you had from Fisher-Yates. The original algorithm says to select an item from the non-shuffled part of the list and you're selecting an item from the entire list.

I had an intuition that this creates a bias and investigated it. Looks like my intuition was correct. I wrote the following python code that does what you do:
```python
def run(n):
    zeros = [0] * n
    ones = [1] * n
    both = zeros + ones
    for i, _ in enumerate(both):
        j = random.randint(0, len(both)-1)
        both[i], both[j] = both[j], both[i]
    return sum(both[:n])
```

If you run `run(x)` with any `x` above 1000 you almost always get a result lower than `x/2`.

I then tested the correct implementation:
```python
def run_fixed(n):
    zeros = [0] * n
    ones = [1] * n
    both = zeros + ones
    for i, _ in enumerate(both):
        j = random.randint(0, i)
        both[i], both[j] = both[j], both[i]
    return sum(both[:n])
```
And this time I get results as expected (about half the times it's above `x/2`).

Comment (noamnelke) [reply p2p/peers.go]: Here are example runs:
```
In [61]: run(1000000)
Out[61]: 477398

In [62]: run(1000000)
Out[62]: 477205

In [63]: run_fixed(1000000)
Out[63]: 499975

In [64]: run_fixed(1000000)
Out[64]: 500306
```

Comment (almogdepaz) [reply p2p/peers.go]: dude you had me in "its already implemented somewhere else"

**Review by noamnelke — APPROVED** [review]

**Pull request merged**